Feature: test async structures naming

We do not check any names in `async` members.

What we need to do:

0. Implement missing support for `async` stuff
1. Add unit tests for `async def` for function names and argument names
2. Add unit tests for `async for` and `async` comprehension variable names
3. Add unit tests for `async with` variable names

Tests are located here: https://github.com/wemake-services/wemake-python-styleguide/tree/master/tests/test_visitors/test_ast/test_general/test_wrong_name

We do not need to test `async` code as much as we test the regular one in case it will share the same logic. We can only write one test case for each violation and good path.

I have parsing error for some special words in BAD_VARIABLE_NAMES (for ex: 'await' and 'async'):
```
>       return compile(source, filename, mode, PyCF_ONLY_AST)
E         File "<unknown>", line 3
E           async for async in []:
E                         ^
E       SyntaxError: invalid syntax
```
How correctly handle these exceptions?

--- a/tests/test_visitors/test_ast/test_general/test_wrong_name/test_function_names.py
+++ b/tests/test_visitors/test_ast/test_general/test_wrong_name/test_function_names.py
@@ -18,6 +18,10 @@ function_bad_name = """
 def {0}(): ...
 """
 
+async_function_bad_name = """
+async def {0}(): ...
+"""
+
 method_bad_name = """
 class Input:
     def {0}(self): ...
@@ -27,6 +31,7 @@ class Input:
 @pytest.mark.parametrize('bad_name', BAD_VARIABLE_NAMES)
 @pytest.mark.parametrize('code', [
     function_bad_name,
+    async_function_bad_name,
     method_bad_name,
 ])
 def test_wrong_function_names(
@@ -44,6 +49,7 @@ def test_wrong_function_names(
 @pytest.mark.parametrize('short_name', string.ascii_letters)
 @pytest.mark.parametrize('code', [
     function_bad_name,
+    async_function_bad_name,
     method_bad_name,
 ])
 def test_too_short_function_names(
@@ -60,6 +66,7 @@ def test_too_short_function_names(
 
 @pytest.mark.parametrize('code', [
     function_bad_name,
+    async_function_bad_name,
     method_bad_name,
 ])
 def test_private_function_names(
@@ -83,6 +90,7 @@ def test_private_function_names(
 ])
 @pytest.mark.parametrize('code', [
     function_bad_name,
+    async_function_bad_name,
     method_bad_name,
 ])
 def test_correct_function_names(

--- a/tests/test_visitors/test_ast/test_general/test_wrong_name/test_variable_names.py
+++ b/tests/test_visitors/test_ast/test_general/test_wrong_name/test_variable_names.py
@@ -31,9 +31,21 @@ for {0} in []:
     print()
 """
 
+async_for_variable_test = """
+async def container():
+    async for {0} in []:
+        print()
+"""
+
 with_variable_test = """
 with open('test.py') as {0}:
     raise ValueError()
+"""
+
+async_with_variable_test = """
+async def container():
+    async with open('test.py') as {0}:
+        raise ValueError()
 """
 
 exception_test = """
@@ -48,7 +60,9 @@ except Exception as {0}:
 @pytest.mark.parametrize('code', [
     variable_test,
     for_variable_test,
+    async_for_variable_test,
     with_variable_test,
+    async_with_variable_test,
     exception_test,
     underscore_variable_test1,
     underscore_variable_test2,
@@ -69,7 +83,9 @@ def test_wrong_variable_names(
 @pytest.mark.parametrize('code', [
     variable_test,
     for_variable_test,
+    async_for_variable_test,
     with_variable_test,
+    async_with_variable_test,
     exception_test,
 ])
 def test_too_short_variable_names(
@@ -87,7 +103,9 @@ def test_too_short_variable_names(
 @pytest.mark.parametrize('code', [
     variable_test,
     for_variable_test,
+    async_for_variable_test,
     with_variable_test,
+    async_with_variable_test,
     exception_test,
 ])
 def test_too_short_variable_names_configured(
@@ -106,7 +124,9 @@ def test_too_short_variable_names_configured(
 @pytest.mark.parametrize('code', [
     variable_test,
     for_variable_test,
+    async_for_variable_test,
     with_variable_test,
+    async_with_variable_test,
     exception_test,
 ])
 def test_private_variable_names(
@@ -125,7 +145,9 @@ def test_private_variable_names(
 @pytest.mark.parametrize('code', [
     variable_test,
     for_variable_test,
+    async_for_variable_test,
     with_variable_test,
+    async_with_variable_test,
     exception_test,
     underscore_variable_test1,
     underscore_variable_test2,

--- a/wemake_python_styleguide/constants.py
+++ b/wemake_python_styleguide/constants.py
@@ -8,7 +8,6 @@ It also contains some exceptions that we allow to use in our codebase.
 """
 
 import re
-import sys
 
 # TODO: use consistent `.` for the `#:` comments
 # TODO: use consistent names: `*_BLACKLIST` and `*_WHITELIST`

--- a/wemake_python_styleguide/constants.py
+++ b/wemake_python_styleguide/constants.py
@@ -51,7 +51,7 @@ BAD_MODULE_METADATA_VARIABLES = frozenset((
     '__about__',
 ))
 
-
+#: List of variable names we forbid to use.
 BAD_VARIABLE_NAMES = frozenset((
     # Meaningless words:
     'data',

--- a/wemake_python_styleguide/constants.py
+++ b/wemake_python_styleguide/constants.py
@@ -53,7 +53,7 @@ BAD_MODULE_METADATA_VARIABLES = frozenset((
 ))
 
 
-_BAD_VARIABLE_NAMES = [
+BAD_VARIABLE_NAMES = frozenset((
     # Meaningless words:
     'data',
     'result',
@@ -86,17 +86,7 @@ _BAD_VARIABLE_NAMES = [
     'foo',
     'bar',
     'baz',
-]
-
-if sys.version_info < (3, 7):  # pragma: no cover
-    _BAD_VARIABLE_NAMES.extend([
-        # Compatibility with `python3.7`:
-        'async',
-        'await',
-    ])
-
-#: List of variable names we forbid to use.
-BAD_VARIABLE_NAMES = frozenset(_BAD_VARIABLE_NAMES)
+))
 
 #: List of magic methods that are forbiden to use.
 BAD_MAGIC_METHODS = frozenset((

--- a/wemake_python_styleguide/types.py
+++ b/wemake_python_styleguide/types.py
@@ -32,6 +32,9 @@ VisitorClass = Type['base.BaseVisitor']
 #: In cases we need to work with both import types:
 AnyImport = Union[ast.Import, ast.ImportFrom]
 
+#: In cases we need to work with both function definitions:
+MethodMembers = Union[ast.FunctionDef, ast.AsyncFunctionDef]
+
 #: Flake8 API format to return error messages:
 CheckResult = Tuple[int, int, str, type]
 

--- a/wemake_python_styleguide/visitors/ast/complexity/counts.py
+++ b/wemake_python_styleguide/visitors/ast/complexity/counts.py
@@ -5,7 +5,7 @@ from collections import defaultdict
 from typing import DefaultDict, Union
 
 from wemake_python_styleguide.logics.functions import is_method
-from wemake_python_styleguide.types import AnyImport
+from wemake_python_styleguide.types import AnyImport, MethodMembers
 from wemake_python_styleguide.violations.complexity import (
     TooManyConditionsViolation,
     TooManyImportsViolation,
@@ -17,7 +17,6 @@ from wemake_python_styleguide.visitors.decorators import alias
 
 ConditionNodes = Union[ast.If, ast.While, ast.IfExp]
 ModuleMembers = Union[ast.AsyncFunctionDef, ast.FunctionDef, ast.ClassDef]
-MethodMembers = Union[ast.FunctionDef, ast.AsyncFunctionDef]
 
 
 @alias('visit_module_members', (

--- a/wemake_python_styleguide/visitors/ast/naming.py
+++ b/wemake_python_styleguide/visitors/ast/naming.py
@@ -28,7 +28,8 @@ from wemake_python_styleguide.visitors.decorators import alias
     'visit_ImportFrom',
     'visit_Import',
 ))
-@alias('visit_FunctionDef', (
+@alias('visit_any_function', (
+    'visit_FunctionDef',
     'visit_AsyncFunctionDef',
 ))
 class WrongNameVisitor(BaseNodeVisitor):
@@ -78,7 +79,7 @@ class WrongNameVisitor(BaseNodeVisitor):
 
         self.generic_visit(node)
 
-    def visit_FunctionDef(self, node: MethodMembers) -> None:
+    def visit_any_function(self, node: MethodMembers) -> None:
         """
         Used to find wrong function and method parameters.
 

--- a/wemake_python_styleguide/visitors/ast/naming.py
+++ b/wemake_python_styleguide/visitors/ast/naming.py
@@ -1,6 +1,7 @@
 # -*- coding: utf-8 -*-
 
 import ast
+from typing import Union
 
 from wemake_python_styleguide.constants import (
     BAD_MODULE_METADATA_VARIABLES,
@@ -27,6 +28,9 @@ from wemake_python_styleguide.visitors.decorators import alias
 @alias('visit_any_import', (
     'visit_ImportFrom',
     'visit_Import',
+))
+@alias('visit_FunctionDef', (
+    'visit_AsyncFunctionDef',
 ))
 class WrongNameVisitor(BaseNodeVisitor):
     """
@@ -75,7 +79,7 @@ class WrongNameVisitor(BaseNodeVisitor):
 
         self.generic_visit(node)
 
-    def visit_FunctionDef(self, node: ast.FunctionDef) -> None:
+    def visit_FunctionDef(self, node: Union[ast.FunctionDef, ast.AsyncFunctionDef]) -> None:
         """
         Used to find wrong function and method parameters.
 

--- a/wemake_python_styleguide/visitors/ast/naming.py
+++ b/wemake_python_styleguide/visitors/ast/naming.py
@@ -1,7 +1,6 @@
 # -*- coding: utf-8 -*-
 
 import ast
-from typing import Union
 
 from wemake_python_styleguide.constants import (
     BAD_MODULE_METADATA_VARIABLES,
@@ -12,7 +11,7 @@ from wemake_python_styleguide.logics.variables import (
     is_too_short_variable_name,
     is_wrong_variable_name,
 )
-from wemake_python_styleguide.types import AnyImport
+from wemake_python_styleguide.types import AnyImport, MethodMembers
 from wemake_python_styleguide.violations.best_practices import (
     WrongModuleMetadataViolation,
 )
@@ -79,10 +78,7 @@ class WrongNameVisitor(BaseNodeVisitor):
 
         self.generic_visit(node)
 
-    def visit_FunctionDef(
-        self,
-        node: Union[ast.FunctionDef, ast.AsyncFunctionDef],
-    ) -> None:
+    def visit_FunctionDef(self, node: MethodMembers) -> None:
         """
         Used to find wrong function and method parameters.
 

--- a/wemake_python_styleguide/visitors/ast/naming.py
+++ b/wemake_python_styleguide/visitors/ast/naming.py
@@ -50,7 +50,7 @@ class WrongNameVisitor(BaseNodeVisitor):
         if is_private_variable(name):
             self.add_violation(PrivateNameViolation(node, text=name))
 
-    def _check_function_signature(self, node: ast.FunctionDef) -> None:
+    def _check_function_signature(self, node: MethodMembers) -> None:
         for arg in node.args.args:
             self._check_name(node, arg.arg)
 

--- a/wemake_python_styleguide/visitors/ast/naming.py
+++ b/wemake_python_styleguide/visitors/ast/naming.py
@@ -79,7 +79,10 @@ class WrongNameVisitor(BaseNodeVisitor):
 
         self.generic_visit(node)
 
-    def visit_FunctionDef(self, node: Union[ast.FunctionDef, ast.AsyncFunctionDef]) -> None:
+    def visit_FunctionDef(
+        self,
+        node: Union[ast.FunctionDef, ast.AsyncFunctionDef],
+    ) -> None:
         """
         Used to find wrong function and method parameters.
 

